File: java/src/main/java/ge/hashbank/baas/gateway/GatewayKeyGenerator.java
package ge.hashbank.baas.gateway;

import java.security.KeyFactory;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.HexFormat;

public final class GatewayKeyGenerator {
    private GatewayKeyGenerator() {
    }

    public static GatewayKeyPair generate() {
        try {
            var generator = KeyPairGenerator.getInstance("Ed25519");
            var keyPair = generator.generateKeyPair();
            var privateKeyPem = toPem("PRIVATE KEY", keyPair.getPrivate().getEncoded());
            var publicKeyPem = toPem("PUBLIC KEY", keyPair.getPublic().getEncoded());
            return new GatewayKeyPair(privateKeyPem, publicKeyPem, deriveKeyId(publicKeyPem));
        } catch (Exception ex) {
            throw new IllegalStateException("Failed to generate Ed25519 key pair", ex);
        }
    }

    public static String deriveKeyId(String publicKeyPem) {
        try {
            var der = parsePem(publicKeyPem, "PUBLIC KEY");
            var digest = MessageDigest.getInstance("SHA-256").digest(der);
            return HexFormat.of().formatHex(digest);
        } catch (Exception ex) {
            throw new IllegalArgumentException("Failed to derive key ID from public key PEM", ex);
        }
    }

    public static PrivateKey importPrivateKey(String privateKeyPem) {
        try {
            var der = parsePem(privateKeyPem, "PRIVATE KEY");
            return KeyFactory.getInstance("Ed25519").generatePrivate(new PKCS8EncodedKeySpec(der));
        } catch (Exception ex) {
            throw new IllegalArgumentException("Invalid Ed25519 PKCS#8 private key PEM", ex);
        }
    }

    public static PublicKey importPublicKey(String publicKeyPem) {
        try {
            var der = parsePem(publicKeyPem, "PUBLIC KEY");
            return KeyFactory.getInstance("Ed25519").generatePublic(new X509EncodedKeySpec(der));
        } catch (Exception ex) {
            throw new IllegalArgumentException("Invalid Ed25519 SPKI public key PEM", ex);
        }
    }

    public static void validateKeyPair(String privateKeyPem, String publicKeyPem) {
        try {
            var payload = "hash-baas-gateway-sdk-key-validation".getBytes(java.nio.charset.StandardCharsets.UTF_8);
            var signer = Signature.getInstance("Ed25519");
            signer.initSign(importPrivateKey(privateKeyPem));
            signer.update(payload);
            var signature = signer.sign();

            var verifier = Signature.getInstance("Ed25519");
            verifier.initVerify(importPublicKey(publicKeyPem));
            verifier.update(payload);
            if (!verifier.verify(signature)) {
                throw new IllegalArgumentException("Private key does not match public key");
            }
        } catch (IllegalArgumentException ex) {
            throw ex;
        } catch (Exception ex) {
            throw new IllegalArgumentException("Failed to validate key pair", ex);
        }
    }

    static byte[] parsePem(String pem, String label) {
        if (pem == null || pem.isBlank()) {
            throw new IllegalArgumentException(label + " PEM is required");
        }
        var base64 = pem
                .replace("-----BEGIN " + label + "-----", "")
                .replace("-----END " + label + "-----", "")
                .replaceAll("\\s", "");
        return Base64.getDecoder().decode(base64);
    }

    private static String toPem(String label, byte[] der) {
        var base64 = Base64.getMimeEncoder(64, "\n".getBytes(java.nio.charset.StandardCharsets.US_ASCII)).encodeToString(der);
        return "-----BEGIN " + label + "-----\n" + base64 + "\n-----END " + label + "-----";
    }
}

File: src/Hash.BaaS.Gateway.Sdk/GatewayKeyGenerator.cs
using System.Security.Cryptography;
using System.Text;
using NSec.Cryptography;

namespace Hash.BaaS.Gateway.Sdk;

/// <summary>
/// Generates Ed25519 key pairs and derives HashBank Gateway key IDs.
/// </summary>
public static class GatewayKeyGenerator
{
    /// <summary>
    /// Generates a new Ed25519 key pair and derives the key ID from the public key.
    /// </summary>
    public static GatewayKeyPair Generate()
    {
        var creationParameters = new KeyCreationParameters
        {
            ExportPolicy = KeyExportPolicies.AllowPlaintextExport
        };

        using var key = Key.Create(SignatureAlgorithm.Ed25519, creationParameters);

        var privateKeyPem = Encoding.UTF8.GetString(key.Export(KeyBlobFormat.PkixPrivateKeyText));
        var publicKeyPem = Encoding.UTF8.GetString(key.PublicKey.Export(KeyBlobFormat.PkixPublicKeyText));
        var keyId = DeriveKeyId(publicKeyPem);

        return new GatewayKeyPair(privateKeyPem, publicKeyPem, keyId);
    }

    /// <summary>
    /// Derives the Gateway key ID as lowercase hex SHA-256(SPKI DER) from an SPKI public key PEM.
    /// </summary>
    public static string DeriveKeyId(string publicKeyPem)
    {
        if (string.IsNullOrWhiteSpace(publicKeyPem))
            throw new ArgumentException("Public key PEM is required.", nameof(publicKeyPem));

        var base64 = publicKeyPem
            .Replace("-----BEGIN PUBLIC KEY-----", string.Empty, StringComparison.Ordinal)
            .Replace("-----END PUBLIC KEY-----", string.Empty, StringComparison.Ordinal)
            .Replace("\r", string.Empty, StringComparison.Ordinal)
            .Replace("\n", string.Empty, StringComparison.Ordinal)
            .Replace("\t", string.Empty, StringComparison.Ordinal)
            .Replace(" ", string.Empty, StringComparison.Ordinal)
            .Trim();

        var spkiDer = Convert.FromBase64String(base64);
        return Convert.ToHexString(SHA256.HashData(spkiDer)).ToLowerInvariant();
    }

    /// <summary>
    /// Validates that a PKCS#8 private key PEM can be imported for Ed25519 signing.
    /// </summary>
    public static void ValidatePrivateKey(string privateKeyPem)
    {
        if (string.IsNullOrWhiteSpace(privateKeyPem))
            throw new ArgumentException("Private key PEM is required.", nameof(privateKeyPem));

        using var _ = Key.Import(
            SignatureAlgorithm.Ed25519,
            Encoding.UTF8.GetBytes(privateKeyPem),
            KeyBlobFormat.PkixPrivateKeyText);
    }

    /// <summary>
    /// Validates that the private key can sign and the public key can verify the same payload.
    /// </summary>
    public static void ValidateKeyPair(string privateKeyPem, string publicKeyPem)
    {
        if (string.IsNullOrWhiteSpace(publicKeyPem))
            throw new ArgumentException("Public key PEM is required.", nameof(publicKeyPem));

        var payload = Encoding.UTF8.GetBytes("hash-baas-gateway-sdk-key-validation");

        using var privateKey = Key.Import(
            SignatureAlgorithm.Ed25519,
            Encoding.UTF8.GetBytes(privateKeyPem),
            KeyBlobFormat.PkixPrivateKeyText);

        var publicKey = PublicKey.Import(
            SignatureAlgorithm.Ed25519,
            Encoding.UTF8.GetBytes(publicKeyPem),
            KeyBlobFormat.PkixPublicKeyText);

        var signature = SignatureAlgorithm.Ed25519.Sign(privateKey, payload);
        if (!SignatureAlgorithm.Ed25519.Verify(publicKey, payload, signature))
            throw new ArgumentException("PrivateKeyPem does not match PublicKeyPem.");
    }
}
